### PR TITLE
feat(ticket#169): Add Search Functionality to Member List (Search by Name & Phone Number)

### PR DESCRIPTION
**Pull request by nadirhussain786**
## Change Request

This PR introduces functionality to the search bar on the Member List page, allowing users to filter members based on their name or phone number.

### Implementation

* Implement a global state using the `useContext API`.
* The global state will store and manage the search query.
* Implement `useDebounce` hook
* Create a custom hook to access the search query from the global state, ensuring error handling.
* Add search functionality  with `Supabase SQL query`
* Add `Loader`  while searching members.

### Resources

image

Screenshot 2024-11-19 at 7 04 00 PM

### CODE REVIEW

**Review by Monfernape — CHANGES_REQUESTED** [review]
@muhammad-bilal-ali-khokhar @nadirhussain786  I don't think this is the right pattern. Two missing things here:
- How about we introduce debouncing here?
- Maybe handle search through DB.
- Pass on query to search params and let server component handle the rest.

File: src/app/attendance/actions/getAttendance.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ import { getMembers } from "@/app/members/actions/getMembers";
 export const getAttendance = async () => {
   const supabase = await getSupabaseClient();
 
-  const membersResponse = await getMembers();
+  const membersResponse = await getMembers("");
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/attendance/actions/getAttendance.ts:10]: @nadirhussain786 why you passing string?

Comment (nadirhussain786) [reply src/app/attendance/actions/getAttendance.ts:10]: @muhammad-bilal-ali-khokhar  Actually, I added the search functionality, which is why I passed an empty string as the payload.

File: src/app/members/components/MemberHeader.tsx [hunk 2]
```diff
@@ -9,17 +9,34 @@ import { usePathname } from "next/navigation";
...
+  const debounceValue = useDebounce(search, 500);
+  const { setSearchValue } = useContext(MemberContext) as MemberContextType;
   const [isSidebarOpen, setSidebarOpen] = useState<boolean>(false);
   const [isTitleHide, setIsTitleHide] = useState<boolean>(false);
 
   const toggleSidebar = () => {
     setSidebarOpen(!isSidebarOpen);
   };
 
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
```
Comment (Monfernape) [inline src/app/members/components/MemberHeader.tsx:32]: Function name is way too generic. Should be something like `handleSearchQueryChange`

File: src/common/Loader/Loader.tsx [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import React from "react";
```
Comment (Monfernape) [inline src/common/Loader/Loader.tsx:1]: File name is weird. Also, this should be in components folder where all components from shadcn reside.

**Review by Monfernape — CHANGES_REQUESTED** [review]
@muhammad-bilal-ali-khokhar @nadirhussain786 approach for search is wrong. We shouldn't use context. Let API be judge of that. Talk to me if I'm not clear enough.

**Review by muhammad-bilal-ali-khokhar — APPROVED** [review]

**Pull request merged**